File: Leetcode/440.py
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        k -= 1
        curr = 1

        while k > 0:
            steps = self.count_steps(n, curr)
            if steps <= k:
                k -= steps
                curr += 1
            else:
                curr *= 10
                k -= 1
        return curr

    @staticmethod
    def count_steps(n: int, curr: int):
        steps = 0
        nxt = curr + 1
        while curr <= n:
            steps += min(n + 1, nxt) - curr
            curr *= 10
            nxt *= 10
        return steps

File: Leetcode/440.go
package Leetcode

func findKthNumber(n int, k int) int {
	curr := 1
	k--

	for {
		if k <= 0 {
			break
		}
		steps := countSteps(n, curr)

		if steps > k {
			curr *= 10
			k--
		} else {
			curr++
			k -= steps
		}
	}

	return curr

}

func countSteps(n, curr int) int {
	steps := 0
	nxt := curr + 1

	for {
		if curr > n {
			break
		}
		steps += min(n+1, nxt) - curr
		curr *= 10
		nxt *= 10
	}

	return steps
}
